src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,238 @@
+            private const int SizeOrLengthOffset = 4;
+            private const int NumberOfRowsOffset = 8;
+
+            internal int Length;
+            private byte[] _rentedBuffer;
+
+            internal CustomDb(int payloadLength)
+            {
+                // Assume that a token happens approximately every 12 bytes.
+                // int estimatedTokens = payloadLength / 12
+                // now acknowledge that the number of bytes we need per token is 12.
+                // So that's just the payload length.
+                //
+                // Add one token's worth of data just because.
+                int initialSize = DbRow.Size + payloadLength;
+
+                // Stick with ArrayPool's rent/return range if it looks feasible.
+                // If it's wrong, we'll just grow and copy as we would if the tokens
+                // were more frequent anyways.
+                const int OneMegabyte = 1024 * 1024;
+                
+                if (initialSize > OneMegabyte && initialSize <= 4 * OneMegabyte)
+                {
+                    initialSize = OneMegabyte;
+                }
+
+                _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialSize);
+                Length = 0;
+            }
+
+            public void Dispose()
+            {
+                if (_rentedBuffer == null)
+                {
+                    return;
+                }
+
+                // The data in this rented buffer only conveys the positions and
+                // lengths of tokens in a document, but no content; so it does not
+                // need to be cleared.
+                ArrayPool<byte>.Shared.Return(_rentedBuffer);
+                _rentedBuffer = null;
+                Length = 0;
+            }
+
+            internal void TrimExcess()
+            {
+                if (Length <= _rentedBuffer.Length / 2)
+                {
+                    byte[] newRent = ArrayPool<byte>.Shared.Rent(Length);
+                    byte[] returnBuf = newRent;
+
+                    if (newRent.Length < _rentedBuffer.Length)
+                    {
+                        Buffer.BlockCopy(_rentedBuffer, 0, newRent, 0, Length);
+                        returnBuf = _rentedBuffer;
+                        _rentedBuffer = newRent;
+                    }
+
+                    // The data in this rented buffer only conveys the positions and
+                    // lengths of tokens in a document, but no content; so it does not
+                    // need to be cleared.
+                    ArrayPool<byte>.Shared.Return(returnBuf);
+                }
+            }
+
+            internal void Append(JsonTokenType tokenType, int startLocation, int length)
+            {
+                // StartArray or StartObject should have length -1, otherwise the length should not be -1.
+                Debug.Assert(
+                    (tokenType == JsonTokenType.StartArray || tokenType == JsonTokenType.StartObject) ==
```

Comment:
nit: That's an odd way to write a condition (at least one I am not familiar with).

Why not use `&&` and explicitly match the comment?

```C#
Debug.Assert((tokenType == JsonTokenType.StartArray || tokenType == JsonTokenType.StartObject) 
&& (length == DbRow.UnknownSize) ||
(tokenType == JsonTokenType.X || tokenType == JsonTokenType.Y || ...) 
&& (length != DbRow.UnknownSize));
```